node/src/components/consensus/highway_core/state.rs

```diff
@@ -915,43 +883,39 @@ impl<C: Context> State<C> {
 
     /// Returns whether the unit with `hash0` sees the one with `hash1` (i.e. `hash0 ≥ hash1`),
     /// and sees `hash1`'s creator as correct.
-    fn _sees_correct(&self, hash0: &C::Hash, hash1: &C::Hash) -> bool {
-        hash0 == hash1 || self.unit(hash0).panorama._sees_correct(self, hash1)
+    pub(crate) fn sees_correct(&self, hash0: &C::Hash, hash1: &C::Hash) -> bool {
+        hash0 == hash1 || self.unit(hash0).panorama.sees_correct(self, hash1)
     }
 
     /// Returns whether the unit with `hash0` sees the one with `hash1` (i.e. `hash0 ≥ hash1`).
-    fn _sees(&self, hash0: &C::Hash, hash1: &C::Hash) -> bool {
-        hash0 == hash1 || self.unit(hash0).panorama._sees(self, hash1)
+    pub(crate) fn sees(&self, hash0: &C::Hash, hash1: &C::Hash) -> bool {
+        hash0 == hash1 || self.unit(hash0).panorama.sees(self, hash1)
     }
 
     // Returns whether the units with `hash0` and `hash1` see each other or are equal.
-    fn _is_compatible(&self, hash0: &C::Hash, hash1: &C::Hash) -> bool {
-        hash0 == hash1
-            || self.unit(hash0).panorama._sees(self, hash1)
-            || self.unit(hash1).panorama._sees(self, hash0)
+    fn is_compatible(&self, hash0: &C::Hash, hash1: &C::Hash) -> bool {
+        hash0 == hash1 || self.unit(hash0).panorama.sees(self, hash1)
     }
 
-    /// Returns the panorama of the confirmation for the leader unit `vhash`.
+    /// Returns the panorama of the confirmation unit for the leader unit `uhash`.
     pub(crate) fn confirmation_panorama(
         &self,
         own_idx: ValidatorIndex,
-        vhash: &C::Hash,
+        uhash: &C::Hash,
     ) -> Panorama<C> {
-        let unit = self.unit(vhash);
-        let mut panorama;
-        // TODO(HWY-167): Confirmation panorama.
-        if let Some(prev_hash) = self.panorama().get(own_idx).correct().cloned() {
-            let own_unit = self.unit(&prev_hash);
-            panorama = unit.panorama.merge(self, &own_unit.panorama);
-            panorama[own_idx] = Observation::Correct(prev_hash);
-        } else {
-            panorama = unit.panorama.clone();
-        }
-        panorama[unit.creator] = Observation::Correct(*vhash);
-        for faulty_v in self.faulty_validators() {
-            panorama[faulty_v] = Observation::Faulty;
+        let mut confirmation_panorama = self.inclusive_panorama(uhash);
+        if confirmation_panorama[own_idx] != self.citable_panorama()[own_idx] {
+            confirmation_panorama = self.citable_panorama().clone();
         }
-        panorama
+        confirmation_panorama
```

Comment:
That still returns the panorama even if the unit isn't citable.

But maybe we should even move that check into `ActiveValidator::new_unit`, too, and make it more general: If the to-be-created vote's panorama is not less or equal than our citable panorama, use the citable panorama instead.

Reply:
We start with `inclusive_panorama` which would have been rejected by the `validate_lnc` in the first place. So it is safe to cite it. If `inlusive_panorama` does not cite our own latest unit then it will be replaced by `citable_panorama` which also is safe to cite. I don't understand why it wouldn't be citable.

Reply:
Wouldn't it fail in [this example](https://github.com/CasperLabs/casper-node/pull/572#discussion_r529416365)?

The danger is that the unit is perfectly valid (satisfies the LNC), cites our previous unit, but can't be added to our existing `citable_panorama`, because we already included another fork in that one (which we haven't cited yet). So we couldn't add our own confirmation unit that we're about to create to `citable_panorama`.

Reply:
OK, it's my fault for not understanding your proposal – why did we change it in the first place then? Previously, when it was returning `Option<Panorama>` all we had to do was add an additional check that the confirmation panorama sees our own latest unit – if not, return `None` or send a witness-like vote.

Reply:
You're right, I mixed a few things there that weren't related, sorry!